### PR TITLE
PLAT-52107: Improve a11y support in ExpandablePicker, DayPicker, and DaySelector

### PR DESCRIPTION
**Pull request by Baekwoo-s**
### Issue Resolved / Feature Added
There's no way to override a11y string on ExpandablePicker, ExpandableItem, ExpandableList, DayPicker

### Resolution
Added a11y props in ExpandablePicker, ExpandableList, DayPicker
- ExpandablePicker
  - checkButtonAriaLabel
  - decrementAriaLabel
  - incrementAriaLabel
  - pickerAriaLabel
- ExpandableList
  - aria-label
  - childrenAriaLabels
- DayPicker
  - childrenAriaLabels
  - everyDayAriaLabel
  - everyWeekdayAriaLabel
  - everyWeekendAriaLabel

ExpandableItem does not add props because it can set the aria-label directly in the child dom.

### Links
PLAT-52107

### Comments
Enact-DCO-1.0-Signed-off-by: Baekwoo Jung (baekwoo.jung@lge.com)

### CODE REVIEW

File: packages/moonstone/CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,9 @@ The following is a curated list of changes in the Enact moonstone module, newest
 - `moonstone/TimePicker` props `hourAccessibilityHint`, `hourAriaLabel`, `meridiemAccessibilityHint`, `meridiemAriaLabel`, `minuteAccessibilityHint`, `minuteAriaLabel` to configure the label set on time pickers
 - `moonstone/VirtualList`, `moonstone/VirtualGridList` and `moonstone/Scroller` props `scrollRightAriaLabel`, `scrollLeftAriaLabel`, `scrollDownAriaLabel`, and `scrollUpAriaLabel` to configure the aria-label set on scroll buttons in the scrollbars
 - `moonstone/Popup` property `closeButtonAriaLabel` to configure the label set on popup close button
+- `moonstone/ExpandablePicker` props `checkButtonAriaLabel`, `decrementAriaLabel`, `incrementAriaLabel`, `pickerAriaLabel` to configure the label set on each button and picker
```
Comment (MikyungKim) [inline packages/moonstone/CHANGELOG.md]: , and `pickerAriaLabel`

File: packages/moonstone/CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,9 @@ The following is a curated list of changes in the Enact moonstone module, newest
...
+- `moonstone/ExpandableList` props `aria-label`, `childrenAriaLabels` to configure the label set on each item
```
Comment (MikyungKim) [inline packages/moonstone/CHANGELOG.md]: `aria-label` and `childrenAriaLabels`

File: packages/moonstone/CHANGELOG.md [hunk 3]
```diff
@@ -13,6 +13,9 @@ The following is a curated list of changes in the Enact moonstone module, newest
...
+- `moonstone/DayPicker` props `childrenAriaLabels`, `everyDayAriaLabel`, `everyWeekdayAriaLabel`, `everyWeekendAriaLabel` to configure the label set on each item
```
Comment (MikyungKim) [inline packages/moonstone/CHANGELOG.md]: same as ExpandablePicker

File: packages/moonstone/DayPicker/DayPicker.js [hunk 4]
```diff
@@ -50,6 +50,14 @@ const DayPickerBase = class extends React.Component {
 		 */
 		title: PropTypes.string.isRequired,
 
+		/**
+		 * Sets the hint string read when focusing the each item.
+		 *
+		 * @type {Array}
+		 * @public
+		 */
+		childrenAriaLabels: PropTypes.array,
```
Comment (MikyungKim) [inline packages/moonstone/DayPicker/DayPicker.js]: How about `dayAriaLabels` for clarity?

File: packages/moonstone/DayPicker/DayPicker.js [hunk 5]
```diff
@@ -198,11 +233,37 @@ const DayPickerBase = class extends React.Component {
...
+	 * @param {Number} selected day type
+	 * @param {String[]} selected day aria labels
+	 * @param {String[]} selected day strings
+	 *
+	 * @returns {String} Aria label for "Every Day", "Every Weekend", "Every Week" or list of days
+	 */
+	getSelectedDayAriaLabel = (type, selectDayAriaLabels, selectDayStrings) => {
+		const
+			{everyDayAriaLabel, everyWeekdayAriaLabel, everyWeekendAriaLabel} = this.props,
+			selected = coerceArray(this.props.selected);
```
Comment (MikyungKim) [inline packages/moonstone/DayPicker/DayPicker.js]: Let's move `this.props.selected` to above line.

File: packages/moonstone/DayPicker/DayPicker.js [hunk 6]
```diff
@@ -198,11 +233,37 @@ const DayPickerBase = class extends React.Component {
 		}
 	}
 
+	/**
+	 * Determines whether it should return ariaLabel passed in prop.
+	 *
+	 * @param {Number} selected day type
+	 * @param {String[]} selected day aria labels
```
Comment (MikyungKim) [inline packages/moonstone/DayPicker/DayPicker.js]: I think this explanation should be `* @param {String[]} Array of day aria labels`.
It's not selected.

File: packages/moonstone/DayPicker/DayPicker.js [hunk 7]
```diff
@@ -198,11 +233,37 @@ const DayPickerBase = class extends React.Component {
...
+	 * @param {String[]} selected day strings
```
Comment (MikyungKim) [inline packages/moonstone/DayPicker/DayPicker.js]: it should be `* @param {String[]} Array of long day strings`

File: packages/moonstone/DayPicker/DayPicker.js [hunk 8]
```diff
@@ -198,11 +233,37 @@ const DayPickerBase = class extends React.Component {
...
+	 *
+	 * @returns {String} Aria label for "Every Day", "Every Weekend", "Every Week" or list of days
```
Comment (MikyungKim) [inline packages/moonstone/DayPicker/DayPicker.js]: list of selected days would makes sense

File: packages/moonstone/DayPicker/DayPicker.js [hunk 9]
```diff
@@ -198,11 +233,37 @@ const DayPickerBase = class extends React.Component {
...
+			return selected.sort().map((dayIndex) => (selectDayAriaLabels && selectDayAriaLabels[dayIndex]) || selectDayStrings[dayIndex]).join(', ');
+		}
+	}
+
 	/**
 	 * Determines whether it should return "Every Day", "Every Weekend", "Every Weekday" or list of
 	 * days for a given selected day type.
 	 *
 	 * @param {Number} selected day type
+	 * @param {String[]} selected day strings
```
Comment (MikyungKim) [inline packages/moonstone/DayPicker/DayPicker.js]: * @param {String[]} Array of long or short day strings

File: packages/moonstone/DayPicker/DayPicker.js [hunk 10]
```diff
@@ -230,21 +291,26 @@ const DayPickerBase = class extends React.Component {
...
-			{title, ...rest} = this.props,
+			{childrenAriaLabels, everyDayAriaLabel, everyWeekdayAriaLabel, everyWeekendAriaLabel, title, ...rest} = this.props,
 			type = this.calcSelectedDayType(this.props.selected),
 			label = this.getSelectedDayString(type, this.shortDayNames);
 
 		delete rest.locale;
 
 		let ariaLabel = null;
-		if (type === SELECTED_DAY_TYPES.SELECTED_DAYS) {
+		if (label &&
```
Comment (MikyungKim) [inline packages/moonstone/DayPicker/DayPicker.js]: Let's simplify this part.

File: packages/moonstone/ExpandableList/ExpandableList.js [hunk 11]
```diff
@@ -271,6 +288,36 @@ const ExpandableListBase = kind({
 	computed: {
 		'aria-multiselectable': ({select}) => select === 'multiple',
 
+		ariaLabel: ({'aria-label': ariaLabel, childrenAriaLabels, label, select, selected, title}) => {
+			if (ariaLabel != null) {
```
Comment (MikyungKim) [inline packages/moonstone/ExpandableList/ExpandableList.js]: Let's refactor this part.

**Comment by MikyungKim** [comment]
Please fix the travis errors.

File: packages/moonstone/CHANGELOG.md (hunk 1 above)
Comment (Baekwoo-s) [reply packages/moonstone/CHANGELOG.md]: I added `and` between last `PROP`

File: packages/moonstone/CHANGELOG.md (hunk 2 above)
Comment (Baekwoo-s) [reply packages/moonstone/CHANGELOG.md]: I added this.

File: packages/moonstone/CHANGELOG.md (hunk 3 above)
Comment (Baekwoo-s) [reply packages/moonstone/CHANGELOG.md]: I added this.

File: packages/moonstone/DayPicker/DayPicker.js (hunk 4 above)
Comment (Baekwoo-s) [reply packages/moonstone/DayPicker/DayPicker.js]: I agree. I have corrected the name.

File: packages/moonstone/DayPicker/DayPicker.js (hunk 7 above)
Comment (Baekwoo-s) [reply packages/moonstone/DayPicker/DayPicker.js]: I agree and will fix.

**Comment by Baekwoo-s** [comment]
I've applied all the above comments, and I've also fixed the triavis error.

**Comment by MikyungKim** [comment]
LGTM! 
@ryanjduffy, please review this PR.

**Comment by ryanjduffy** [comment]
The PR was in pretty good shape but I made some changes driven by code maintainability and upstream changes to DaySelectorDecorator.

Summary of changes:

* Removed the `childrenAriaProps` from ExpandableList and DayPicker. For ExpandableList, that can be supported currently using an object array so there's minimal value in adding it as a first-class prop on the component. For DayPicker, I made a judgement call that it make little sense to allow custom aria labels for the day strings. If that requirement comes up later, we can consider it then.
```
<ExpandableList title="...">
  {[
    {children: 'first', 'aria-label': 'first label'},
    {children: 'second', 'aria-label': 'second label'}
  ]}
</ExpandableList>
```
* Batch of refactoring to DaySelectorDecorator to consolidate logic and adequately support the ARIA requirements
* Added `aria-label`s to DaySelector's items to include full day names rather than abbreviated.

**Pull request merged**